src/rgw/rgw_cr_rados.cc

```diff
@@ -917,15 +917,24 @@ int RGWContinuousLeaseCR::operate(const DoutPrefixProvider *dpp)
     return set_cr_done();
   }
   reenter(this) {
+    last_renew_try_time = ceph::real_time::clock::now();
     while (!going_down) {
       yield call(new RGWSimpleRadosLockCR(async_rados, store, obj, lock_name, cookie, interval));
+      current_time = ceph::real_time::clock::now();
+      if (current_time - last_renew_try_time > interval_tolerance) {
+        // renewal should happen between 50%-90% of interval
+        ldout(store->ctx(), 1) << *this << ": WARNING: did not renew lock " << obj << ":" << lock_name << ": within 90\% of interval. " << 
+          (current_time - last_renew_try_time) << " > " << interval_tolerance << dendl;
```

Comment:
i'm not crazy about warning at 90% unless we can offer specific advice to prevent the user from hitting this 100% latency barrier

i'm more concerned that this 100% barrier isn't an actual error. your change to `is_locked()` helps, but we can't guarantee that `is_locked()` will be called promptly. another rgw could still acquire and drop the lock in the meantime, and we'd log a warning here but sync would continue on like nothing happened

our lock requests are sent with `set_may_renew(true)` (see https://github.com/ceph/ceph/blob/f5c21acf/src/rgw/rgw_cr_rados.cc#L217) which means the request can either acquire or renew a lock. if `RGWContinuousLeaseCR` were to send its renewals with `set_must_renew(true)` instead, the OSD would reject renewal requests if the previous lock had expired

`RGWContinuousLeaseCR` would see that as an `EBUSY` error on renewal. can we put a good log message there instead, with some explanation like "high latency on the pool {obj.pool}"?

Reply:
> i'm not crazy about warning at 90% unless we can offer specific advice to prevent the user from hitting this 100% latency barrier
> 
> i'm more concerned that this 100% barrier isn't an actual error. your change to `is_locked()` helps, but we can't guarantee that `is_locked()` will be called promptly. another rgw could still acquire and drop the lock in the meantime, and we'd log a warning here but sync would continue on like nothing happened
> 

i can remove this warning. it does not say that lock was not successful, just that it was late.
my idea was to correlate that message with sync issues

> our lock requests are sent with `set_may_renew(true)` (see https://github.com/ceph/ceph/blob/f5c21acf/src/rgw/rgw_cr_rados.cc#L217) which means the request can either acquire or renew a lock. if `RGWContinuousLeaseCR` were to send its renewals with `set_must_renew(true)` instead, the OSD would reject renewal requests if the previous lock had expired
> 
> `RGWContinuousLeaseCR` would see that as an `EBUSY` error on renewal. can we put a good log message there instead, with some explanation like "high latency on the pool {obj.pool}"?

but the same continuous lease CR is used for the initial acquisition of the lock and its renewal.
maybe we can add an API to that class that allows switching between the modes?

Reply:
> but the same continuous lease CR is used for the initial acquisition of the lock and its renewal.
> maybe we can add an API to that class that allows switching between the modes?

exactly, right. i opened https://github.com/ceph/ceph/pull/47809 (but haven't compiled or tested it) that extends `RGWSimpleRadosLockCR` to support this. so `RGWContinuousLeaseCR` could have a `rados::cls::lock::Lock` member variable, and use `set_may_renew()` for the initial lock request and `set_must_renew()` afterwards

Reply:
> i can remove this warning. it does not say that lock was not successful, just that it was late. my idea was to correlate that message with sync issues

yeah, i think the proposed error message meets that same goal. even if `is_locked()` sees that the lock probably expired, it will use `drain_all()` to wait for `RGWContinuousLeaseCR` to receive this lock response and print its error message before returning the `-ECANCELED` error up the call stack

Reply:
i'm happy to take up the rest of this work in https://github.com/ceph/ceph/pull/47809, we can merge this warning as is